### PR TITLE
Implement IIndexQueueProcessor support

### PR DESCRIPTION
**Pull request by ericof**
This is a major refactoring, removing `collective.elasticsearch.es` and `collective.elasticsearch.hooks` files and reducing the number of calls to elastic search server.

- Implement IIndexQueueProcessor support
- Refactor ElasticSearchCatalog methods into ElasticSearchManager object
- Breaking: Remove collective.elasticsearch.es
- Breaking: Remove collective.elasticsearch.hooks
- Refactor moveObjectsByDelta to reduce the number of calls to ElasticSearch
- Reindex operations update on the catalog send only updated index to ElasticSearch

### CODE REVIEW

**Review by wesleybl — CHANGES_REQUESTED** [review]
I did the following steps:

- Before installing the product, I created and published a Document with the title "Página"
- I installed the product.
- Converted rebuild the catalog.
- I did a search for the word "Página".

I received the error below:

```bash
2022-09-20 19:48:02 ERROR [Zope.SiteErrorLog:35][waitress-0] TypeError: http://localhost:8080/Plone/@@search
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 167, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 376, in publish_module
  Module ZPublisher.WSGIPublisher, line 271, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 68, in call_object
  Module zope.browserpage.simpleviewclass, line 41, in __call__
  Module Products.Five.browser.pagetemplatefile, line 126, in __call__
  Module Products.Five.browser.pagetemplatefile, line 58, in __call__
  Module zope.pagetemplate.pagetemplate, line 133, in pt_render
  Module Products.PageTemplates.engine, line 365, in __call__
  Module z3c.pt.pagetemplate, line 176, in render
  Module chameleon.zpt.template, line 302, in render
  Module chameleon.template, line 215, in render
  Module chameleon.utils, line 53, in raise_with_traceback
  Module chameleon.template, line 192, in render
  Module 13941a397226e26770f236f36f6221dc, line 3865, in render
  Module ad2563f39a8e59cc044a43efa1485321, line 921, in render_master
  Module ad2563f39a8e59cc044a43efa1485321, line 1581, in render_content
  Module 13941a397226e26770f236f36f6221dc, line 3687, in __fill_main
  Module 13941a397226e26770f236f36f6221dc, line 427, in render_search_results
  Module Products.PageTemplates.engine, line 74, in __call__
  Module plone.batching.batch, line 161, in __getitem__
  Module plone.app.contentlisting.contentlisting, line 30, in __getitem__
TypeError: ('Could not adapt', None, <InterfaceClass plone.app.contentlisting.interfaces.IContentListingObject>)

 - Expression: "batch"
 - Filename:   ... e-packages/Products/CMFPlone/browser/templates/search.pt
 - Location:   (line 203: col 44)
 - Source:     <tal:results repeat="item batch">
                                         ^^^^^
 - Expression: "context/@@main_template/macros/master"
 - Filename:   ... e-packages/Products/CMFPlone/browser/templates/search.pt
 - Location:   (line 6: col 23)
 - Source:     ... tal:use-macro="context/@@main_template/macros/master"
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 - Arguments:  template: <Products.Five.browser.pagetemplatefile.ViewPageTemplateFile object at 0x7f016375c970>
               options: {}
               args: ()
               nothing: None
               modules: <Products.PageTemplates.ZRPythonExpr._SecureModuleImporter object at 0x7f016e967d00>
               request: <WSGIRequest, URL=http://localhost:8080/Plone/@@search>
               view: <Products.Five.browser.metaconfigure.SimpleViewClass from /home/user/git/collective.elasticsearch/lib/python3.10/site-packages/Products/CMFPlone/browser/templates/search.pt object at 0x7f01608ea800>
               context: <PloneSite at /Plone>
               views: <Products.Five.browser.pagetemplatefile.ViewMapper object at 0x7f0160c4a620>
               here: <PloneSite at /Plone>
               container: <PloneSite at /Plone>
               root: <Application at >
               traverse_subpath: []
               user: <PropertiedUser 'admin'>
               default: <DEFAULT>
               repeat: <Products.PageTemplates.engine.RepeatDictWrapper object at 0x7f016015d000>
               loop: {'type': <Products.PageTemplates.engine.RepeatItem object at 0x7f01601157e0>, 'item': <Products.PageTemplates.engine.RepeatItem object at 0x7f01607355a0>}
               target_language: None
               translate: <function BaseTemplate.render.<locals>.translate at 0x7f0153d7e5f0>
               macroname: 'master'
               attrs: {}
```

**Comment by ericof** [comment]
@wesleybl Which Plone and Python (3.10) versions?
@andreclimaco and I found the same issue and I'm trying to figure it out why the content was not in the catalog anymore

**Comment by wesleybl** [comment]
Plone 6 with Python 3.10. It's even missing Python 3.10 in the classifier. Also missing entry in Changes about Python 3 support.

File: src/collective/elasticsearch/patches/__init__.py [hunk 1]
```diff
@@ -1,71 +1,74 @@
...
 
 
 def unrestrictedSearchResults(self, REQUEST=None, **kw):
-    es = ElasticSearchCatalog(self)
-    return es.searchResults(REQUEST, check_perms=False, **kw)
+    manager = ElasticSearchManager()
+    return manager.search_results(REQUEST, check_perms=False, **kw)
 
 
 def safeSearchResults(self, REQUEST=None, **kw):
```
Comment (wesleybl) [inline src/collective/elasticsearch/patches/__init__.py:16]: Plone's `searchResults` method calls [processQueue](https://github.com/plone/Products.CMFPlone/blob/811621d057c755db97a428c9722702761a59e6c8/Products/CMFPlone/CatalogTool.py#L401). So you need to call it in the patch.

File: src/collective/elasticsearch/manager.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        @param query_params: Parameters to pass to the search method
+            'stored_fields': the list of fields to get from stored source
+        """
+        factory = BrainFactory(self.catalog._catalog)
+        result = ElasticResult(self, query, **query_params)
+        return LazyMap(factory, result, result.count)
+
+    def search_results(self, request=None, check_perms=False, **kw):
+        # Make sure any pending index tasks have been processed
+        processQueue()
```
Comment (wesleybl) [inline src/collective/elasticsearch/manager.py:169]: Try removing commits from search tests, to see how they behave now.

File: src/collective/elasticsearch/patches/__init__.py (hunk 1 above)
Comment (ericof) [reply src/collective/elasticsearch/patches/__init__.py:16]: Done!

File: src/collective/elasticsearch/manager.py (hunk 2 above)
Comment (ericof) [reply src/collective/elasticsearch/manager.py:169]: > I did the following steps:
> 
>     * Before installing the product, I created and published a Document with the title "Página"
> 
>     * I installed the product.
> 
>     * Converted rebuild the catalog.
> 
>     * I did a search for the word "Página".
> 

Could you retry this? @wesleybl

**Comment by ericof** [comment]
@wesleybl I've added tests for:

- [X] Index deleted on portal_catalog
- [X] Object in ES but not available on portal_catalog but existing on ZODB
- [X] Object in ES but not available on portal_catalog and not available on ZODB

**Comment by ericof** [comment]
I did the following steps:

    * Before installing the product, I created and published a Document with the title "Página"

    * I installed the product.

    * Converted rebuild the catalog.

    * I did a search for the word "Página".

Could you retry this? @wesleybl

File: src/collective/elasticsearch/tests/test_search.py [hunk 3]
```diff
@@ -69,180 +59,283 @@ def test_date_index_query(self):
...
         )
-        self.assertEqual(len(el_results), 120)
-        self.assertEqual(el_results[0].getId, "0000page")
-        self.assertEqual(el_results[-1].getId, "0119page")
-        self.assertEqual(el_results[-50].getId, "0070page")
-        self.assertEqual(el_results[-55].getId, "0065page")
-        self.assertEqual(el_results[-100].getId, "0020page")
+        self.commit(wait=1)
+        path = "/".join(self.event.getPhysicalPath())
+        catalog.uncatalogObject(path)
```
Comment (wesleybl) [inline src/collective/elasticsearch/tests/test_search.py:282]: Isn't it risky to keep the registry in Elastic, when Plone tells you to uncatalog an object? I believe that when someone does this, it is for the content not to appear in searches. What's the idea here?

Comment (ericof) [reply src/collective/elasticsearch/tests/test_search.py:282]: This test is to reproduce a situation we had with a client.
For some reason, there were cases where the object was not on Catalog anymore but still existed on ElasticSearch.
There is also the case when the object does not exist on the database anymore but still exists on ElasticSearch.

**Comment by wesleybl** [comment]
> I did the following steps:
>
>* Before installing the product, I created and published a Document with the title "Página"
>
>* I installed the product.
>
>* Converted rebuild the catalog.
>
>* I did a search for the word "Página".
>Could you retry this? @wesleybl

This worked now!

File: src/collective/elasticsearch/tests/test_search.py (hunk 3 above)
Comment (wesleybl) [reply src/collective/elasticsearch/tests/test_search.py:282]: But expected is uncatalogObject remove de record of Elastic. I think you have to use another method, created just for the test, which removes the Elastic record, to simulate this situation.

File: src/collective/elasticsearch/manager.py (hunk 2 above)
Comment (wesleybl) [reply src/collective/elasticsearch/manager.py:176]: Still have to try this.

**Review by wesleybl — CHANGES_REQUESTED** [review]
Add a Changes entry to support Python 3 versions.

Add Python 3.10 to classifiers

**Comment by wesleybl** [comment]
When I do a `Clear and Rebuild` on Catatog, I see the message:

```
Catalog Rebuilt
Total time: 0.24891090393066406
Total CPU time: 0.02017041300000244
```
in the log, but not in the ZMI interface. I get the `204 No Content` status of the request.

When I do a `Clear and Rebuild` in Catatog of a portal that is in an instance that has the `collective.elasticsearch` in the eggs but it is not installed for that portal, I see requests to Elastic in the log. I don't think this is a serious thing, since it's hard to have two portals in one instance, one with the product installed and one without.

File: src/collective/elasticsearch/tests/test_search.py (hunk 3 above)
Comment (wesleybl) [reply src/collective/elasticsearch/tests/test_search.py:282]: You can even do a test in which `catalog.uncatalogObject(path)` is called for a record, and verify that it no longer appears in the search.

Comment (ericof) [reply src/collective/elasticsearch/tests/test_search.py:282]: The use case we have is: An object in the database, for some reason, is not in portal_catalog anymore **but** still exists on ElasticSearch. The ZCatalog.uncatalogObject is the closest method we have to reproducing this (as it is a low-level call that removes the object from the catalog but not from the ZODB or ElasticSearch).

**Comment by ericof** [comment]
@wesleybl Added checks to avoid using Elastic if this product is not installed and enabled

File: src/collective/elasticsearch/manager.py (hunk 2 above)
Comment (ericof) [reply src/collective/elasticsearch/manager.py:188]: Tried to do this but some tests with fail. I will add an issue to refactor all tests, but not for this release.

File: src/collective/elasticsearch/patches/__init__.py [hunk 4]
```diff
@@ -1,71 +1,80 @@
...
+    return method(REQUEST, check_perms=False, **kw)
 
 
 def safeSearchResults(self, REQUEST=None, **kw):
-    es = ElasticSearchCatalog(self)
-    return es.searchResults(REQUEST, check_perms=True, **kw)
+    manager = ElasticSearchManager()
+    method = self._old_unrestrictedSearchResults
+    if manager.active:
+        method = manager.search_results
```
Comment (wesleybl) [inline src/collective/elasticsearch/patches/__init__.py:23]:
```suggestion
    if manager.active:
        method = manager.search_results
    else:
        method = self._old_unrestrictedSearchResults
```

File: src/collective/elasticsearch/tests/test_search.py (hunk 3 above)
Comment (wesleybl) [reply src/collective/elasticsearch/tests/test_search.py:282]: I think you can do the following:

- Patch `uncatalogObject` to remove the record from Elastic.
- Save original `uncatalogObject`.
- Use the original `uncatalogObject` in the test to simulate your situation.

File: src/collective/elasticsearch/patches/__init__.py [hunk 5]
```diff
@@ -1,71 +1,80 @@
...
+from zope.interface import noLongerProvides
 
 
 def unrestrictedSearchResults(self, REQUEST=None, **kw):
-    es = ElasticSearchCatalog(self)
-    return es.searchResults(REQUEST, check_perms=False, **kw)
+    manager = ElasticSearchManager()
+    method = self._old_unrestrictedSearchResults
+    if manager.active:
+        method = manager.search_results
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/elasticsearch/patches/__init__.py:15]: Well, would be one less assignment. But ok.

**Review by wesleybl — CHANGES_REQUESTED** [review]
I still think there is the issue of Clear and Rebuild not displaying message and uncatalog not remove the record of Elastic.

File: src/collective/elasticsearch/tests/test_search.py (hunk 3 above)
Comment (wesleybl) [inline src/collective/elasticsearch/tests/test_search.py:282]: > Just double checking, uncatalog does remove the record from ElasticSearch

I see that in this line the `uncatalogObject` is called with the path of the Event. But in the test_search_results below, the search finds the Event:

https://github.com/collective/collective.elasticsearch/blob/ca4f83b73ecbe4635018accdc0bcfac5db33cc43/src/collective/elasticsearch/tests/test_search.py#L286-L288

What am I missing?

Comment (ericof) [reply src/collective/elasticsearch/tests/test_search.py:282]: @wesleybl This is the expected behavior. There is an object that is not on portal_catalog but still is in ES, so the result should not break during access to the result.
The uncatalogObject is used to simulate this scenario

Comment (ericof) [reply src/collective/elasticsearch/tests/test_search.py:282]: Before that, accessing el_results[0] would return None and break plone.api

File: src/collective/elasticsearch/patches/__init__.py (hunk 4 above)
Comment (ericof) [reply src/collective/elasticsearch/patches/__init__.py]: Changed this to be a ternary :-)

**Comment by ericof** [comment]
@wesleybl "Clear and Rebuild" should be working now

File: src/collective/elasticsearch/tests/test_search.py (hunk 3 above)
Comment (wesleybl) [reply src/collective/elasticsearch/tests/test_search.py:282]: > This is the expected behavior.

I don't think this is the expected behavior. When calling `uncatalogObject`, I expect the record to be removed from Elastic. Even in the master and previus versions it is like this. See:

https://github.com/collective/collective.elasticsearch/blob/0375accc03093c81f16355f773b53b83f317cb57/src/collective/elasticsearch/patches/__init__.py#L13-L15

But you changed this behavior in this PR.

Probably, the situation that occurred of the record being in Elastic and not being in the catalog was due to some error. But to simulate this problem in the test, you are making every call to `uncatalogObject` "generate this error".

I think the error simulation should only be done in the test, not in every `uncatalogObject` call, which can occur outside of the test. In a comment above, I gave a suggestion on how to simulate the error only in the test, and keeping the `uncatalogObject` removing record from Elastic.

Comment (ericof) [reply src/collective/elasticsearch/tests/test_search.py:282]: @wesleybl, now I understand where is our "misunderstanding.":

tl;dr: Removing an object from the catalog should remove the object from Elastic Search using `QueueProcessor.unindex`. The call in the test is to simulate a case where the removal in Elastic Search failed.

Uncataloging the object now happens in [QueueProcessor](https://github.com/collective/collective.elasticsearch/blob/use-queue-processor/src/collective/elasticsearch/queueprocessor.py#L116), and this is called by [CMFCore](https://github.com/zopefoundation/Products.CMFCore/blob/master/src/Products/CMFCore/indexing.py#L228).

Portal catalog also implements un-indexing using a [QueueProcessor](https://github.com/zopefoundation/Products.CMFCore/blob/master/src/Products/CMFCore/indexing.py#L54), and in there you can see the data being removed from the catalog.

What I added to that test is a call to [Catalog.unindexObject](https://github.com/zopefoundation/Products.ZCatalog/blob/master/src/Products/ZCatalog/Catalog.py#L380) that will remove the object from the catalog **without** triggering any additional event or call. 

The idea of the test is to simulate a scenario where:

- User deletes an object on the site
- Transaction happens
- QueueProcessor for the Portal Catalog is effective
- QueueProcessor for ElasticSearch fails to erase the record
- A new query happens, and the *bad* record is returned

Comment (wesleybl) [reply src/collective/elasticsearch/tests/test_search.py:282]: > What I added to that test is a call to [Catalog.unindexObject](https://github.com/zopefoundation/Products.ZCatalog/blob/master/src/Products/ZCatalog/Catalog.py#L380) that will remove the object from the catalog without triggering any additional event or call.

@ericof  about using the `unindexObject` method test us ok. I understood!

But what I still find strange is when I go to the Catalog tab of portal_catalog and order a record to be removed. It is removed from portal_catalog but still in Elastic. In this situation I expect the registration to be removed from Elastic. This action calls the `uncatalog_object` method, which you remove patch in this PR. To see:

https://github.com/zopefoundation/Products.ZCatalog/blob/6885451bc4f4313dba970c4880e610f922e0f13b/src/Products/ZCatalog/ZCatalog.py#L238

Comment (ericof) [reply src/collective/elasticsearch/tests/test_search.py:282]: @wesleybl After sleeping on it, I realized the call via ZMI would not trigger QueueProcessor (as expected), so I'm adding the patch back.

Comment (ericof) [reply src/collective/elasticsearch/tests/test_search.py]: @wesleybl Fixed, and added tests
